packages/adminrouter/extra/src/README.md

```diff
@@ -31,6 +31,229 @@ Use `make check-api-docs` to validate that the YAML and HTML files are up to dat
+remove all open directories from the Open repository before applying EE repository on
+top of it.
+
+This is not a bulletproof solution for preventing code
+duplication (developers can simply start putting copies of code to both
+`open/` and `ee/` directories) but it makes it easier to reuse the
+code and encourages good behaviours.
+
+### Lua code deduplication
+It is not strictly required to provide the same level of flexibility for Lua code
+as it is the case for Nginx includes and thus it's possible to simplify the code
+a bit. It is sufficient to just differentiate Lua code basing on the repository flavour.
+Both agents and masters can share the same Lua code.
+
+There are two possible reasons that may be preventing Lua code from being
+shared:
+* the same code is executed but with different call arguments. An example to
+  this may be `auth.validate_jwt_or_exit()` function. In the Open it takes no arguments,
+  in EE it takes more than one.
+* the code differs between EE and open but shares some common libraries/functions.
+  Great example for this is `auth.check_acl_or_exit()` which internally, among
+  many other things, calls argument-less `auth.check_jwt_or_exit()`.
+
+In order to address these issues, a couple of patterns were selected:
+ * modules which have flavour-specific function arguments export argument-less
+   functions for the NGINX configuration. They translate the original call into
+   a call with correct arguments. This approach requires splitting the module into
+   `ee.lua|open.lua|common.lua` parts which is described in next bullet point.
+   Depending on the flavour, either `ee.lua` or `open.lua` is imported and correct
+   argument-less function is used. This approach also enables to share some of
+   the NGINX `location` blocks between flavours - `location` code is the same,
+   even though the Lua code used by the block differs.
+ * some libraries/functions are structured in a way that extracts parts common for
+   both flavours and `ee/open` parts that are included only in EE and Open repos
+   respectively. An example of this approach is auth library which is splitted
+   into three parts:
+    * `lib/auth/common.lua` - present only in Open repository, with functions
+      shared by both EE and Open code.
+    * `lib/auth/ee.lua` - present only in EE repository, with functions specific
+      to EE that use boilerplate from lib/auth/common.lua.
+    * `lib/auth/open.lua` - as above but for Open repository.
+   `init_by_lua` OpenResty call in `includes/http/(open|ee)/common.conf` imports
+   `auth.open` or `auth.ee` modules as auth respectively.  The module is registered in
+   the global namespace so all other Lua code uses it. This approach also allows
+   for some degree of code separation enforcement as Open lua libs are removed
+   during EE repository apply.
+
+A special case of "same code path, but with different call arguments"
+problem is when a module needs to be initialized differently/passed
```

Comment:
Is there a way a module can be initialised (s in British English) differently without it being passed different arguments? If not, remove the first case.